src/compiler/crystal/codegen/primitives.cr

```diff
@@ -293,10 +293,9 @@ class Crystal::CodeGenVisitor
   private def float32_upper_bound(max_value)
     {% begin %}
       case max_value
-      when Int32, UInt32, Int64, UInt64 {% unless flag?(:win32) %}, Int128, UInt128 {% end %}
-        # TODO: use `prev_float` once 1.2.0 is out
-        # TODO: windows lacks definitions for `__floattisf` and `__floatuntisf`
-        max_value.class.new(max_value.to_f32.unsafe_as(Int32).pred.unsafe_as(Float32))
+      when Int32, UInt32, Int64, UInt64 {% unless flag?(:win32) %}, Int128 {% end %}
```

Comment:
UInt128 missing on purpose?

Reply:
Actually it is correct to omit `UInt128` even outside `win32`, because `UInt128::MAX.to_f32` overflows. This can be refactored to show that (especially after #11437).